pkg/graveler/graveler.go

```diff
@@ -1597,45 +1597,50 @@ func CommitExists(ctx context.Context, repositoryID RepositoryID, commitID Commi
 }
 
 func (g *KVGraveler) AddCommitToBranchHead(ctx context.Context, repositoryID RepositoryID, branchID BranchID, commit Commit) (CommitID, error) {
-	// TODO(issue 3569) - kv implementation optimistic lock commit
-	res, err := g.db.branchLocker.MetadataUpdater(ctx, repositoryID, branchID, func() (interface{}, error) {
-		// parentCommitID should always match the HEAD of the branch.
-		// Empty parentCommitID matches first commit of the branch.
-		parentCommitID, err := validateCommitParent(ctx, repositoryID, commit, g.RefManager)
-		if err != nil {
-			return nil, err
-		}
+	repo, err := g.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return "", err
+	}
 
-		branch, err := g.db.GetBranch(ctx, repositoryID, branchID)
-		if err != nil {
-			return nil, err
-		}
-		if branch.CommitID != parentCommitID {
-			return nil, ErrCommitNotHeadBranch
-		}
+	// parentCommitID should always match the HEAD of the branch.
+	// Empty parentCommitID matches first commit of the branch.
+	parentCommitID, err := validateCommitParent(ctx, repositoryID, commit, g.RefManager)
+	if err != nil {
+		return "", err
+	}
 
-		// check if commit already exists.
-		commitID := CommitID(ident.NewHexAddressProvider().ContentAddress(commit))
-		if exists, err := CommitExists(ctx, repositoryID, commitID, g.RefManager); err != nil {
-			return nil, err
-		} else if exists {
-			return commitID, nil
-		}
+	// verify access to meta range
+	ok, err := g.CommittedManager.Exists(ctx, repo.StorageNamespace, commit.MetaRangeID)
+	if err != nil {
+		return "", fmt.Errorf("commit missing meta range %s: %w", commit.MetaRangeID, err)
+	}
+	if !ok {
+		return "", fmt.Errorf("%w: %s", ErrMetaRangeNotFound, commit.MetaRangeID)
+	}
 
-		commitID, err = g.addCommitNoLock(ctx, repositoryID, commit)
-		if err != nil {
-			return nil, fmt.Errorf("adding commit: %w", err)
-		}
-		_, err = g.UpdateBranch(ctx, repositoryID, branchID, Ref(commitID))
-		if err != nil {
-			return nil, err
+	// add commit and/or use the commit id of the one we found
+	commitID, err := g.RefManager.AddCommit(ctx, repositoryID, commit)
+	if err != nil && !errors.Is(err, kv.ErrPredicateFailed) {
+		return "", fmt.Errorf("adding commit: %w", err)
+	}
+
+	err = g.RefManager.BranchUpdate(ctx, repositoryID, branchID, func(branch *Branch) (*Branch, error) {
+		if branch.CommitID != parentCommitID {
+			return nil, ErrCommitNotHeadBranch
 		}
-		return commitID, nil
+		return &Branch{
+			CommitID:     commitID,
+			StagingToken: branch.StagingToken,
+			SealedTokens: branch.SealedTokens,
```

Comment:
Shouldn't we check that these are not empty?

Reply:
You are right, I missed the part I removed the empty check - there is no way to lock the branch. So checking the sealed tokens will work, but the listing can be an issue.